### PR TITLE
[Svelte] Improve typescript types + minor fixes

### PR DESCRIPTION
**Pull request by jamesst20**
@reinink 

I have rebased all my commits in my repo (https://github.com/jamesst20/inertia) to include the latest changes and afterwards I have taken all my packages/svelte files there to overwrite the ones in the current repo. I have stripped down all unnecessary changes and all new features as well.

This PR only improve the types and fix some possible regression that I will comment in a few seconds in the diff

Edit: @reinink  it's ready. i'm done adding comments :)

### CODE REVIEW

File: packages/svelte/src/lib/components/App.svelte
Comment (jamesst20) [inline packages/svelte/src/lib/components/App.svelte:5]: Does nothing

File: packages/svelte/src/lib/components/Render.svelte
Comment (jamesst20) [inline packages/svelte/src/lib/components/Render.svelte:34]: Prior to the typescript merge we had

```
$: {
    if (prevComponent !== component) {
      key = Date.now()
      prevComponent = component
    }
  }
```

and with the typescript merge we were now having

```
function updateKey(component: InertiaComponentType) {
    if (prev !== component) {
      prev = component
      key = new Date().getTime()
  }
}


$: updateKey(component)
```

While it looks like to be the same, it ain't. Everything between ${...} is reactive and is able to trigger state update unlike when it's wrapped into a method

File: packages/svelte/src/lib/components/SSR.svelte
Comment (jamesst20) [inline packages/svelte/src/lib/components/SSR.svelte:3]: No changes here. Exporting the types to use in createInertiaApp for the SSR instead of casting to any for better typing.

File: packages/svelte/src/lib/useForm.ts
Comment (jamesst20) [inline packages/svelte/src/lib/useForm.ts:177]: This is the real inertia@core signature

File: packages/svelte/src/lib/types.ts
Comment (jamesst20) [inline packages/svelte/src/lib/types.ts:9]: I left `props?: PageProps` but i'm honestly not sure it's actually used at all. I had it deleted in my branch.

Looking here: https://inertiajs.com/pages#default-layouts the return value looks like

```
return { default: page.default, layout: page.layout || Layout }
```

and without a layout it would be 

```
return { default: page.default }
// which is same as
return page
```

File: packages/svelte/tsconfig.json
Comment (joetannenbaum) [inline packages/svelte/tsconfig.json:1]: Could you provide some clarification as to why these items were removed?

File: packages/svelte/src/lib/useForm.ts
Comment (joetannenbaum) [inline packages/svelte/src/lib/useForm.ts:123]: I'm still getting some TS errors when the `ts-ignore` is removed, are you getting those as well?

File: packages/svelte/src/lib/createInertiaApp.ts
Comment (joetannenbaum) [inline packages/svelte/src/lib/createInertiaApp.ts:83]: Instead of the additional `isServer` check here, should we just move this above the check above and remove the second check?

File: packages/svelte/src/lib/types.ts
Comment (joetannenbaum) [inline packages/svelte/src/lib/types.ts:4]: This should remain singular if I'm understanding correctly, yes? `ComponentResolver`, since it resolves a single component?

File: packages/svelte/src/lib/types.ts
Comment (joetannenbaum) [inline packages/svelte/src/lib/types.ts:6]: This should also be singular, yes? `ResolvedComponent`?

File: packages/svelte/src/lib/types.ts
Comment (joetannenbaum) [reply packages/svelte/src/lib/types.ts:9]: It looks like `props` is indeed not used from what I can tell, we may as well remove it.

**Review by joetannenbaum — COMMENTED** [review]
Hi @jamesst20, thanks for this PR! @reinink asked me to give it a review, I have just a couple of items that need clarification whenever you have a moment.

File: packages/svelte/src/lib/types.ts
Comment (jamesst20) [reply packages/svelte/src/lib/types.ts:6]: @joetannenbaum  It would be plural if the layout prop is filled. I do not mind making them singular as it d

File: packages/svelte/src/lib/types.ts
Comment (jamesst20) [reply packages/svelte/src/lib/types.ts:4]: That depends on your answer below about the `ResolvedComponents` :)

File: packages/svelte/src/lib/createInertiaApp.ts
Comment (jamesst20) [reply packages/svelte/src/lib/createInertiaApp.ts:83]: The condition used to be present, it was only removed when the typescript types were merged : https://github.com/inertiajs/inertia/commit/c01b56e5bc094fd5e5e3d1200590969208ba1bf2

We indeed can avoid the condition, however this is more alike of how vue2, vue3 and react adapter works. It's also very clear that the below code is only intended for SSR only

File: packages/svelte/src/lib/useForm.ts
Comment (jamesst20) [reply packages/svelte/src/lib/useForm.ts:123]: Capture d’écran, le 2024-06-26 à 17 02 14

No I don't have any kind of errors. Maybe you have out of dates nodes_modules this can happen easily

From the inertia folder root
```
find . -name 'node_modules' -type d -prune -exec rm -rf '{}' +
find . -name 'dist' -type d -prune -exec rm -rf '{}' +
npm install
cd packages/core && npm run build
cd ../svelte && npm run build
```

Restart your IDE (VSCode or other) and I believe you should be fine too?

I will add extract types just in case

File: packages/svelte/tsconfig.json
Comment (jamesst20) [reply packages/svelte/tsconfig.json:1]: The configuration extends from `"extends": "./.svelte-kit/tsconfig.json",` and this is more inline with the default configuration of a new svelte-kit project. I don't have any kind of objection to add it back. This was just added in the recent commit https://github.com/inertiajs/inertia/commit/c01b56e5bc094fd5e5e3d1200590969208ba1bf2

File: packages/svelte/src/lib/types.ts
Comment (jamesst20) [reply packages/svelte/src/lib/types.ts]: Done ! 👍

File: packages/svelte/src/lib/types.ts
Comment (jamesst20) [reply packages/svelte/src/lib/types.ts]: It's done ! :)

File: packages/svelte/src/lib/types.ts
Comment (jamesst20) [reply packages/svelte/src/lib/types.ts]: It's done! :)

File: packages/svelte/tsconfig.json
Comment (jamesst20) [reply packages/svelte/tsconfig.json:1]: I undid the changes as it really doesn't matter, in the end both way works and it's not worth stalling the PR any longer for this :)

**Comment by jamesst20** [comment]
@joetannenbaum  @reinink 

What would be the next step?

**Comment by jamesst20** [comment]
> @joetannenbaum @reinink
> 
> What would be the next step?

Hey guys @joetannenbaum  @reinink  it's me again bothering you 😆  let me know if I can be of any help

**Comment by joetannenbaum** [comment]
Hey @jamesst20! Sorry for the delay here. This is on my list to review again tomorrow, I'll update you then.

File: packages/svelte/src/lib/useForm.ts
Comment (joetannenbaum) [reply packages/svelte/src/lib/useForm.ts]: Ok, checked it out fresh and looks ok, not sure what was up. Thanks!

File: packages/svelte/src/lib/createInertiaApp.ts
Comment (joetannenbaum) [reply packages/svelte/src/lib/createInertiaApp.ts:83]: The other adapters do the `isServer` check in order to run `setupProgress`, for my money I think it's more readable if we move the `isServer` check up which makes the `if` block shorter and easier to parse, then just return otherwise.

**Comment by kevin82222** [comment]
@joetannenbaum @reinink @jamesst20 Have anything to do more about this?
I looking forward to using inertia@svelte with TypeScript in my project~
Thanks everyone for this pr works!

File: packages/svelte/src/lib/createInertiaApp.ts
Comment (jamesst20) [reply packages/svelte/src/lib/createInertiaApp.ts:52]: Done!

**Comment by jamesst20** [comment]
@joetannenbaum I applied the styling suggestion and rebased on the latest commit too

> @joetannenbaum @reinink @jamesst20 Have anything to do more about this? I looking forward to using inertia@svelte with TypeScript in my project~ Thanks everyone for this pr works!

Not really, I have been running my own typescript fork for few months without troubles

**Comment by jamesst20** [comment]
@reinink  Could you please merge? :)

**Review by pedroborges — CHANGES_REQUESTED** [review]
This is great work @jamesst20, thank you! Please see my comments.

After installing `@types/lodash` I got different type errors. I worked on a fix to those error and will open a PR to your branch. Please review it when you get a chance.

File: packages/svelte/src/lib/components/Render.svelte
Comment (jamesst20) [reply packages/svelte/src/lib/components/Render.svelte]: Done

**Comment by jamesst20** [comment]
@pedroborges  All done! Nice addition to remove all the duplicated calls and checks for the data type being function or object :)

Ready to merge?

**Comment by pedroborges** [comment]
> Ready to merge?

@jamesst20 please install `@types/lodash`. I didn't include that in my PR but only fixed the type errors, sorry for the confusion.

After that, I'd like to see if @joetannenbaum wants to review this once again. Then we will be good to merge.

Thanks again @jamesst20!

**Comment by jamesst20** [comment]
> > Ready to merge?
> 
> @jamesst20 please install `@types/lodash`. I didn't include that in my PR but only fixed the type errors, sorry for the confusion.
> 
> After that, I'd like to see if @joetannenbaum wants to review this once again. Then we will be good to merge.
> 
> Thanks again @jamesst20!

Done!

Screenshot 2024-08-11 at 12 44 13 AM


@joetannenbaum  @pedroborges  @reinink  Let me know if this can be merged

**Review by pedroborges — APPROVED** [review]
Thanks @jamesst20! Just tested locally and it looks good 👏

Let's just wait for @joetannenbaum and @reinink before we merge it 😉

**Comment by jamesst20** [comment]
> Thanks @jamesst20! Just tested locally and it looks good 👏
> 
> Let's just wait for @joetannenbaum and @reinink before we merge it 😉

Could we proceed? It doesn't look like anyone is looking into it?

**Comment by joekrump** [comment]
If others are required for review before this can be merged I suggest adding them as reviewers on the PR so they will be more likely to be aware that they are blocking if that is the case.

**Comment by jamesst20** [comment]
It appears to me the project Inertia is dead and there are no more reasons for me to keep my work open and maintain it.

I am closing all my opened PRs which includes https://github.com/inertiajs/inertia/pull/1864 https://github.com/inertiajs/inertia/pull/1872 https://github.com/inertiajs/inertia/pull/1873 https://github.com/inertiajs/inertia/pull/1874 https://github.com/inertiajs/inertia/pull/1875 https://github.com/inertiajs/inertia/pull/1881

I will be maintaining unofficially my own fork of this library so my projects using Svelte 5 can still live until I find or create an appropriate replacement for this library.

Thanks to everyone who spent time looking at my PRs.

**Comment by kevin82222** [comment]
@jamesst20 It's really unfortunate to hear that the Inertia project might no longer be actively maintained. I understand why you decided to close your PRs, but it's still a pity, given all the valuable contributions you've made.
I'm particularly interested in the unofficial version of the library that you'll be maintaining, especially since I am using Svelte 5 in my projects. Could you please let me know how I can access your fork? It would be a huge help in keeping my projects running smoothly.
Thank you again for all your efforts and for continuing to support the community with your unofficial fork.

**Comment by jamesst20** [comment]
> @jamesst20 It's really unfortunate to hear that the Inertia project might no longer be actively maintained. I understand why you decided to close your PRs, but it's still a pity, given all the valuable contributions you've made. I'm particularly interested in the unofficial version of the library that you'll be maintaining, especially since I am using Svelte 5 in my projects. Could you please let me know how I can access your fork? It would be a huge help in keeping my projects running smoothly. Thank you again for all your efforts and for continuing to support the community with your unofficial fork.

With pleasure :) I'm glad I am not the only one interested into it :)

My fork is here https://github.com/jamesst20/inertia/
 
The main branch is for Svelte 5 and there is also a backup branch that has a Svelte 5 separate adapter. I would recommend you use the main branch since it will be the one I keep up to date but I only completed yesterday. The old branch have months of testing but it's mostly the same but forked from the current typescript implementation instead :)

Backup branch : https://www.npmjs.com/package/@jamesst20/inertia-svelte5

Main branch : https://www.npmjs.com/package/@jamesst20/inertia-svelte

Also use https://www.npmjs.com/package/@jamesst20/inertia-core

**Comment by reinink** [comment]
> maintain



> It appears to me the project Inertia is dead and there are no more reasons for me to keep my work open and maintain it.
> 
> I am closing all my opened PRs which includes #1864 #1872 #1873 #1874 #1875 #1881
> 
> I will be maintaining unofficially my own fork of this library so my projects using Svelte 5 can still live until I find or create an appropriate replacement for this library.
> 
> Thanks to everyone who spent time looking at my PRs.

Bummer dude 😕

We've been working *really* hard on Inertia.js v2.0 (as you can see in [Taylor's Laracon keynote](https://www.youtube.com/watch?v=AwWepVU5uWM)), so the project is far from dead — we're just in a bit of an awkward position between major versions.

I apologize that your work has not received more attention, that's partly my fault for not having enough time to review PRs recently. I don't work on Inertia.js full time like other open source maintainers (ie. Laravel, Livewire), so for me it actually requires time out of my personal life. Because of that things just take longer, as much as I wish it didn't.

Thanks for your contributions either way.

**Comment by pedroborges** [comment]
@jamesst20, I totally get where you're coming from, and I'm really sorry for the frustration you've been feeling. We can definitely do better here. I haven't been as involved as I wished. I was away from the project for the last two years, but I'm back now and really grateful for everyone who's kept contributing in the meantime 🙇

The team has been heads down working on Inertia.js v2, especially leading up to the Laracon announcement last week. There are a lot of cool new features coming, and I'm excited to be more involved again!

Just last Friday I mentioned internally that getting #1881 and #1872 reviewed and approved was my top priority this week since I'm blocked by them too. I'd really appreciate it if you could reopen these two PRs. After that, I'll be focusing on fixing #1670 and adding all the new v2 stuff to the Svelte adapter so we can launch it with the other adapters next month.

Thanks again for all your hard work and contributions. It really means a lot!

**Comment by kevin82222** [comment]
I appreciate the clarification, and I apologize for my earlier comment — it seems I misunderstood the situation. I had been following this PR because I need the functionality it provides, and I mistakenly thought the project might not be actively maintained.
It's great to hear that Inertia.js is actively being worked on and that v2.0 is in progress. I can imagine how challenging it must be to balance open-source work with personal life, so I completely understand the delays.
I'm excited to see the new features in the upcoming version, and I'll continue to follow the project closely. Thanks to the entire team for your hard work and dedication to this project!

**Comment by pedroborges** [comment]
#1963 has been merged and will be tagged in the next release 🥳

**Pull request closed without merging**
